examples/albef/data/vqa_datamodules.py

```diff
@@ -0,0 +1,191 @@
+from examples.albef.data.vqa_dataset import VQADataset
+from pytorch_lightning import LightningDataModule
+from torch import Tensor
+from torch.nn.utils.rnn import pad_sequence
+from torch.utils.data import DataLoader, DistributedSampler
+
+
+class VQADataModule(LightningDataModule):
+    """
+    The Data Module for Visual Question Answering task.
+
+    Args:
+        train_files (List[str]): The paths to training json files.
+        test_files (List[str]): The paths to testing json files.
+        answer_list (str): The path to the answers list.
+        vqa_root (str): The path to vqa data directory.
+        vg_root (str): The path to vg data directory.
+    """
+
+    def __init__(
+        self,
+        train_files: List[str],
+        test_files: List[str],
+        answer_list: str,
+        vqa_root: str,
+        vg_root: str,
+    ) -> None:
+        super().__init__()
+        self.train_dataset = VQADataset(
+            train_files,
+            vqa_root,
+            vg_root,
+            ALBEFTransform(is_train=True),
+            split="train",
+        )
+
+        self.test_dataset = VQADataset(
+            test_files,
+            vqa_root,
+            vg_root,
+            ALBEFTransform(is_train=False),
+            split="test",
+            answer_list=answer_list,
+        )
+
+    def _get_sampler(
+        self,
+        dataset: VQADataset,
+        shuffle: bool,
+        is_distributed: bool,
+        num_tasks: int,
+        global_rank: int,
+    ) -> Union[None, DistributedSampler]:
+        if not is_distributed:
+            return None
+
+        return DistributedSampler(
+            dataset, num_replicas=num_tasks, rank=global_rank, shuffle=shuffle
+        )
+
+    def train_dataloader(
+        self,
+        batch_size: int,
+        num_workers: int,
+        is_distributed: bool = False,
+        num_tasks: int = 0,
+        global_rank: int = 0,
+    ) -> DataLoader:
+        """
+        DataLoader Outputs:
+            images (Tensor): Tensor of shape (B, C, W, H) of image inputs.
+            questions (Tensor): Tensor of shape (B, L) of question inputs.
+            question_atts (Tensor): Tensor of shape (B, L) of question attention mask.
+            answers (Tensor): Tensor of shape (N, M) of answer inputs.
+                N >= B because a vqa sample can have multiple answers.
+            answer_atts (Tensor): Tensor of shape (N, M) of answer attention mask.
+            weights (Tensor): Tensor of shape (N) of answer weights.
+            ans_lengths (List[int]): List of length B and sum N where
+                ans_lengths[i] = number of answers for images[i] and questions[i].
+        """
+        sampler = self._get_sampler(
+            dataset=self.train_dataset,
+            shuffle=True,
+            is_distributed=is_distributed,
+            num_tasks=num_tasks,
+            global_rank=global_rank,
+        )
+        shuffle = sampler is None
```

Comment:
not essential for now, but just note that this may cause problems later on if you want to have deterministic behavior